lib/internal/streams/async_iterator.js

```diff
@@ -0,0 +1,100 @@
+'use strict';
+
+const { promisify } = require('util');
+
+class Item {
+  constructor(value, done) {
+    this.done = done;
+    this.value = value;
+  }
+}
+
+function readAndResolve(iter) {
+  if (iter.lastResolve !== null) {
+    const data = iter.stream.read();
+    // we defer if data is null
+    // we can be expecting either 'end' or
+    // 'error'
+    if (data !== null) {
+      iter.lastResolve(new Item(data, false));
+      iter.lastResolve = null;
+      iter.lastReject = null;
+    }
+  }
+}
+
+class ReadableAsyncIterator {
+  constructor(stream) {
+    this.stream = stream;
+    this.lastResolve = null;
+    this.lastReject = null;
+    this.error = null;
+    this.ended = false;
```

Comment:
Could we store all of these in a way that's not publicly accessible? Whenever anything is made unintentionally public, we usually regret it later. 😞

Reply:
I'm actually considering a `ReadableAsyncInterator` a __private__ object, meaning it's consumed by `for await`, and it's not really user facing. Should we treat it as user-facing? I will replace those with symbols then.

Reply:
Right, but it can be returned by `readable[Symbol.asyncIterator]` so it's not truly private. I think the fact that it's really easy to get it and there might be legitimate use cases for it, makes me uneasy about exposing these props publicly.

Reply:
Regarding the "instance creation" comment above (can't comment on it for some reason) - the scary part isn't creating the objects it's the GC afterwards :)